sway-lib-std/src/inputs.sw

```diff
@@ -13,112 +17,273 @@ use ::tx::{
 // These are based on the old spec (before
 // https://github.com/FuelLabs/fuel-specs/pull/400) because that's what's
 // currently implemented in `fuel-core`, `fuel-asm`, and `fuel-tx. They should
 // eventually be updated.
-// const GTF_INPUT_MESSAGE_WITNESS_INDEX = 0x11A;
-// const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
-// const GTF_INPUT_MESSAGE_PREDICATE_LENGTH = 0x11C;
-// const GTF_INPUT_MESSAGE_PREDICATE_DATA_LENGTH = 0x11D;
-// const GTF_INPUT_MESSAGE_DATA = 0x11E;
-// const GTF_INPUT_MESSAGE_PREDICATE = 0x11F;
-const GTF_INPUT_MESSAGE_PREDICATE_DATA = 0x120;
+pub const GTF_INPUT_MESSAGE_WITNESS_INDEX = 0x11A;
+pub const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
+pub const GTF_INPUT_MESSAGE_PREDICATE_LENGTH = 0x11C;
+pub const GTF_INPUT_MESSAGE_PREDICATE_DATA_LENGTH = 0x11D;
+pub const GTF_INPUT_MESSAGE_DATA = 0x11E;
+pub const GTF_INPUT_MESSAGE_PREDICATE = 0x11F;
+pub const GTF_INPUT_MESSAGE_PREDICATE_DATA = 0x120;
 
 pub enum Input {
     Coin: (),
     Contract: (),
     Message: (),
 }
 
+impl Eq for Input {
+    fn eq(self, other: Self) -> bool {
+        match (self, other) {
+            (Input::Coin, Input::Coin) => true,
+            (Input::Contract, Input::Contract) => true,
+            (Input::Message, Input::Message) => true,
+            _ => false,
+        }
+    }
+}
+
+////////////////////////////////////////
+// General Inputs
+////////////////////////////////////////
 /// Get the type of the input at `index`.
 pub fn input_type(index: u64) -> Input {
-    let type = __gtf::<u8>(index, GTF_INPUT_TYPE);
-    match type {
+    match __gtf::<u8>(index, GTF_INPUT_TYPE) {
         0u8 => Input::Coin,
         1u8 => Input::Contract,
         2u8 => Input::Message,
         _ => revert(0),
     }
 }
 
-/// for either tx type (transaction-script or transaction-create).
+/// Get the transaction inputs count
+pub fn input_count() -> u8 {
+    match tx_type() {
+        Transaction::Script => __gtf::<u8>(0, GTF_SCRIPT_INPUTS_COUNT),
+        Transaction::Create => __gtf::<u8>(0, GTF_CREATE_INPUTS_COUNT),
+    }
+}
+
+/// Get the pointer of the input at `index`.
 pub fn input_pointer(index: u64) -> u64 {
-    let type = tx_type();
-    match type {
+    match tx_type() {
         Transaction::Script => __gtf::<u64>(index, GTF_SCRIPT_INPUT_AT_INDEX),
         Transaction::Create => __gtf::<u64>(index, GTF_CREATE_INPUT_AT_INDEX),
     }
 }
 
-/// If the input's type is `InputCoin` the owner as an Option::Some(owner).
+/// Get amount field from input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the amount as an `Option::Some(u64)`.
+/// Otherwise, returns `Option::None`.
+pub fn input_amount(index: u64) -> Option<u64> {
+    match input_type(index) {
+        Input::Coin => Option::Some(__gtf::<u64>(index, GTF_INPUT_COIN_AMOUNT)),
+        Input::Message => Option::Some(__gtf::<u64>(index, GTF_INPUT_MESSAGE_AMOUNT)),
+        Input::Contract => return Option::None,
```

Comment:
How come some of these have a `return` and other `match` expressions do not? Is it needed in these cases?